org.eclipse.xtext/src/org/eclipse/xtext/resource/containers/StateBasedContainer.java

```diff
@@ -44,19 +46,24 @@ public StateBasedContainer(IResourceDescriptions descriptions, IContainerState s
 	
 	@Override
 	protected Iterable<IEObjectDescription> filterByURI(Iterable<IEObjectDescription> unfiltered) {
-		return Iterables.filter(unfiltered, new Predicate<IEObjectDescription>() {
+		Predicate<IEObjectDescription> predicate = getStateContainsPredicate(input -> input.getEObjectURI().trimFragment());
+		return Iterables.filter(unfiltered, predicate);
+	}
+	
+	private <T> Predicate<T> getStateContainsPredicate(Function<T, URI> uriProvider) {
+		return new Predicate<T>() {
 			private Collection<URI> contents = null;
 
 			@Override
-			public boolean apply(IEObjectDescription input) {
+			public boolean apply(T input) {
 				if(contents == null) {
```

Comment:
Looks good to me, though maybe `state.contains(..)` is the better option here, while we are on it. @cdietrich what's your take on this?

Reply:
the problem is that i have no idea how to test this. i dont have a usecase for that. we ususally dont have 100s of projects so the change is irrelevent

Reply:
2 projects / jars with 10000 files in the first and 1 file in the second should allow to observe an improvement when looking for objects by type in the second container.

Reply:
finally i did a little test for state.contains vs. contents.contains. on a cleanbuild of my test workspace the build took 220s with state.contains vs 210s with the cached contents.contains.

so it seems there is a benefit of getting the contents and to the check on them

Reply:
:+1: